draft-ietf-mls-protocol.md

```diff
@@ -4872,27 +4885,73 @@ handle extensible fields:
+  GroupInfo message as invalid and not join the group.
 
-Note that the latter two requirements mean that all MLS extensions are
-mandatory, in the sense that an extension in use by the group MUST be supported
-by all members of the group.
+Note that the latter two requirements mean that all MLS GroupContext extensions
+are mandatory, in the sense that an extension in use by the group MUST be
+supported by all members of the group.
 
 The parameters of a group may be changed by sending a GroupContextExtensions
-proposal to enable additional extensions, or by reinitializing the group as
-described in {{reinitialization}}.
+proposal to enable additional extensions ({{groupcontextextensions}}), or
+by reinitializing the group ({{reinitialization}}).
+
+## GREASE
+
+As described in {{extensions}}, clients are required to ignore unknown values
+for certain parameters.  To help ensure that other clients implement this
+behavior, a client can follow the “Generate Random Extensions And Sustain
+Extensibility” or GREASE approach described in {{?RFC8701}}.  In the context of
+MLS, this means that a client creating a LeafNode or Welcome message includes
+random values in certain fields, which should be ignored by a
+correctly-implemented client processing the message.  A client that incorrectly
+rejects unknown code points will fail to process such a message, providing a
+signal to its implementer that the client needs to be fixed.
+
+When generating the following fields, an MLS client SHOULD include a random
+selection of values chosen from these GREASE values:
+
+* `LeafNode.capabilities.versions`
+* `LeafNode.capabilities.ciphersuites`
+* `LeafNode.capabilities.extensions`
+* `LeafNode.capabilities.proposals`
+* `LeafNode.capabilities.credentials`
+* `KeyPackage.extensions`
+* `GroupInfo.extensions`
+
+For the KeyPackage and GroupInfo extensions, the `extension_data` for GREASE
+extensions MAY have any contents selected by the sender, since they will be
+ignored by a correctly-operating receiver.  For example, a senders might
+populate these extensions with a randomly-sized amount of random data.
+
+A set of values reserved for GREASE have been registered in the various
+registries in {{iana-considerations}}.  This prevents conflict between GREASE
+and real future values.  The following values are reserved in each registry:
+`0x0A0A`, `0x1A1A`, `0x2A2A`, `0x3A3A`, `0x4A4A`, `0x5A5A`, `0x6A6A`, `0x7A7A`,
+`0x8A8A`, `0x9A9A`, `0xAAAA`, `0xBABA`, `0xCACA`, `0xDADA`, and `0xEAEA`.  (The
+value `0xFAFA` falls within the private use range.). These values MUST only
+appear in the fields listed above, and not, for example, in the `proposal_type`
+field of a Proposal.  Clients MUST NOT implement any special processing rules 
+for how to handle these values when receiving them, since this negates their
+utility for detecting extensibility failures.
+GREASE values MUST be handled using normal logic for processing unsupported
+values.
```

Comment:
@DavidSchinazi is there a standard way to deal with the possibility that you'll accidentally negotiate a GREASE code point?

The best that comes to my mind is something like the following:
> When comparing lists of capabilities to identify mutually-supported capabilities, clients MUST represent their own capabilities with a list containing only the capabilities actually supported, without any GREASE values.  In other words, lists including GREASE values are only sent to other clients; representations of a client's own capabilities MUST NOT contain GREASE values.

(The latter part of which is basically what @Bren2010 says.)

Reply:
It's never possible to negotiate a GREASE value because the recipients ignore them. Conceptually the algorithm for matching is "take the set of received values, and the set of supported values, and use the intersection set to decide which one to negotiate" - and that's regardless of GREASE. Perhaps:

> GREASE values are never added to the list of locally supported capabilities when negotiating capabilities.

Reply:
Yeah, the problem arises when you look at your own KeyPackage to remind yourself of your capabilities, and accidentally get GREASE in your eyes.  (Sorry)  Which seems a little outlandish, but seems worth mentioning since "won't we accidentally negotiate GREASE values?" also came up in my mind and I had to talk myself back.

Reply:
```suggestion
GREASE values MUST be handled using normal logic for processing unsupported
values.  When comparing lists of capabilities to identify mutually-supported
capabilities, clients MUST represent their own capabilities with a list
containing only the capabilities actually supported, without any GREASE values.
In other words, lists including GREASE values are only sent to other clients;
representations of a client's own capabilities MUST NOT contain GREASE values.
```